File: core/src/main/java/com/emarsys/core/util/ImageUtils.java
package com.emarsys.core.util;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.webkit.URLUtil;

import com.emarsys.core.device.DeviceInfo;

import java.io.File;

public class ImageUtils {

    public static Bitmap loadBitmap(Context context, String imageUrl) {
        Bitmap result = null;
        if (ImageUtils.isImageUrlValid(imageUrl)) {
            String fileUrl = downloadImage(context, imageUrl);
            result = loadBitmap(fileUrl, Integer.MAX_VALUE);
            if (fileUrl != null && isRemoteUrl(imageUrl)) {
                FileUtils.delete(fileUrl);
            }
        }
        return result;
    }

    public static Bitmap loadOptimizedBitmap(Context context, String imageUrl, DeviceInfo deviceInfo) {
        Assert.notNull(context, "Context must not be null!");
        Assert.notNull(deviceInfo, "DeviceInfo must not be null!");

        Bitmap result = null;
        if (ImageUtils.isImageUrlValid(imageUrl)) {
            String fileUrl = downloadImage(context, imageUrl);
            result = loadBitmap(fileUrl, deviceInfo.getDisplayMetrics().widthPixels);
            if (fileUrl != null && isRemoteUrl(imageUrl)) {
                FileUtils.delete(fileUrl);
            }
        }
        return result;
    }

    private static boolean isImageUrlValid(String imageUrl) {
        boolean result = true;
        if (imageUrl == null) {
            result = false;
        } else if (!URLUtil.isHttpsUrl(imageUrl)) {
            result = new File(imageUrl).exists();
        }
        return result;
    }

    private static String downloadImage(Context context, String imageUrl) {
        String fileUrl = imageUrl;
        if (isRemoteUrl(imageUrl)) {
            fileUrl = FileUtils.download(context, imageUrl);
        }
        return fileUrl;
    }

    private static boolean isRemoteUrl(String imageUrl) {
        return URLUtil.isHttpsUrl(imageUrl);
    }

    private static Bitmap loadBitmap(String imageFileUrl, int width) {
        final BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(imageFileUrl, options);

        options.inSampleSize = calculateInSampleSize(options, width);

        options.inJustDecodeBounds = false;
        return BitmapFactory.decodeFile(imageFileUrl, options);
    }

    static int calculateInSampleSize(BitmapFactory.Options options, int reqWidth) {
        final int width = options.outWidth;
        int inSampleSize = 1;
        while (reqWidth <= width / inSampleSize) {
            inSampleSize *= 2;
        }
        return inSampleSize;
    }

}

File: core/src/main/java/com/emarsys/core/util/ImageUtils.kt
package com.emarsys.core.util

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.webkit.URLUtil
import com.emarsys.core.device.DeviceInfo
import com.emarsys.core.util.log.Logger
import com.emarsys.core.util.log.entry.CrashLog
import java.io.File

object ImageUtils {
    private const val RETRY_COUNT = 3

    fun loadOptimizedBitmap(fileDownloader: FileDownloader, imageUrl: String?, deviceInfo: DeviceInfo): Bitmap? {
        var result: Bitmap? = null
        if (isImageUrlValid(imageUrl)) {
            val fileUrl = downloadImage(fileDownloader, imageUrl)
            if (fileUrl != null) {
                result = loadBitmap(fileUrl, deviceInfo.displayMetrics.widthPixels)
                if (isRemoteUrl(imageUrl)) {
                    fileDownloader.delete(fileUrl)
                }
            }
        }
        return result
    }

    private fun isImageUrlValid(imageUrl: String?): Boolean {
        var result = true
        if (imageUrl == null) {
            result = false
        } else if (!URLUtil.isHttpsUrl(imageUrl)) {
            result = File(imageUrl).exists()
        }
        return result
    }

    private fun downloadImage(fileDownloader: FileDownloader, imageUrl: String?): String? {
        var fileUrl = imageUrl
        if (isRemoteUrl(imageUrl) && imageUrl != null) {
            fileUrl = fileDownloader.download(imageUrl, RETRY_COUNT)
        }
        return fileUrl
    }

    private fun isRemoteUrl(imageUrl: String?): Boolean {
        return URLUtil.isHttpsUrl(imageUrl)
    }

    private fun loadBitmap(imageFileUrl: String, width: Int): Bitmap? = try {
        val options = BitmapFactory.Options().apply {
            inJustDecodeBounds = true
        }
        BitmapFactory.decodeFile(imageFileUrl, options)
        options.inSampleSize = calculateInSampleSize(options, width)
        options.inJustDecodeBounds = false
        BitmapFactory.decodeFile(imageFileUrl, options)
    } catch (exception:Exception) {
        Logger.error(CrashLog(exception))
        null
    }

    fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int): Int {
        val width = options.outWidth
        var inSampleSize = 1
        while (reqWidth <= width / inSampleSize) {
            inSampleSize *= 2
        }
        return inSampleSize
    }
}